src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala

```diff
@@ -75,49 +128,127 @@ object RockSlicer {
-        val sortedNonPersistent = nonPersistentJoints.sortWith(_.centerX < _.centerX)
-        val indices = loadBalanceIndices(sortedNonPersistent.indices).take(numNonPersistent)
-        val seedNonPersistent = indices.collect(sortedNonPersistent)
-        val remainingNonPersistent = sortedNonPersistent.diff(seedNonPersistent)
-        (persistentJoints ++ seedNonPersistent, remainingNonPersistent)
-    }
-  }
-
-  private def loadBalanceIndices(indices: Seq[Int]): Seq[Int] = {
-    def loadBalanceIndicesImpl(queue: mutable.PriorityQueue[Seq[Int]]): Seq[Int] = {
-      if (queue.isEmpty) {
-        Nil
-      } else {
-        val s = queue.dequeue()
-        val (median, left, right) = removeMedian(s)
-        if (left.nonEmpty) {
-          queue.enqueue(left)
+  /**
+    * Recursive function that merges sequence of blocks containing processor joints
+    * along matching processor joints, elliminating false blocks caused by division of 
+    * input rock volume into equal volumes by load balancer
+    * @param processorBlocks Sequence of blocks containing one or more processor joints each
+    * @param mergedBlocks Sequence of blocks that have had processor joints removed. Initial
+    *                     input is empty sequence
+    * @param origin Origin that all blocks and their faces will be referenced to for consistency
+    * @return Sequence of blocks with all processor joints removed by merging input blocks along
+    *         matching processor joints. Redundant faces and duplicate blocks are already removed
+    *         in this list, so it is not necessary to perform these computations on the returned
+    *         sequence.
+    */
+  @tailrec
+  def mergeBlocks(processorBlocks: Seq[Block], mergedBlocks: Seq[Block],
+                  origin: (Double, Double, Double)): Seq[Block] = {
+    val reconstructedBlocks = (processorBlocks map { block =>
+      val currentBlock = Block(origin, processorBlocks.head.updateFaces(origin))
+      val updatedBlock = Block(origin, block.updateFaces(origin))
+      if (!currentBlock.approximateEquals(updatedBlock)) {
+        val sharedProcFaces = compareProcessorBlocks(currentBlock, updatedBlock)
+        if (sharedProcFaces.nonEmpty) {
+          val currentFaces = currentBlock.faces.diff(sharedProcFaces)
+          val updatedFaces = updatedBlock.faces.diff(sharedProcFaces)
+          val allFaces = currentFaces ++ updatedFaces
+          // Check for any actual shared faces between blocks - if these exists blocks
+          // should not be merged since there is a real joint seperating the blocks
+          val nonSharedFaces =
+            allFaces.foldLeft(Seq.empty[Face]) { (unique, current) =>
+              if (!unique.exists(compareSharedFaces(current, _))) {
+                current +: unique
+              } else {
+                unique
+              }
+            }
+          if (allFaces.diff(nonSharedFaces).isEmpty) {
```

Comment:
How about returning `Some(Block ...)` in the `if` and `None` in the `else`? Then just flatten at the end instead of collecting.